core/templates/dev/head/pages/topic_editor/subtopics_editor/SubtopicsListTabDirective.js

```diff
@@ -34,90 +34,155 @@ oppia.directive('subtopicsListTab', [
+                  $scope.closeSubtopicTitleEditor = function() {
+                    $scope.subtopicTitleEditorIsShown = false;
+                  };
 
-          $scope.$on(EVENT_SUBTOPIC_PAGE_LOADED, function() {
-            $scope.subtopicPage = TopicEditorStateService.getSubtopicPage();
-            $scope.subtopicDisplayed = true;
-            $scope.subtopicEditorIsShown = false;
-            $scope.htmlData = $scope.subtopicPage.getHtmlData();
-          });
+                  $scope.openPreviewSubtopicPage = function(htmlData) {
+                    $scope.subtopicEditorIsShown = false;
+                    $scope.htmlData = htmlData;
+                  };
 
-          $scope.openSubtopicTitleEditor = function() {
-            $scope.subtopicTitleEditorIsShown = true;
+                  $scope.closePreviewSubtopicPage = function(previewHtmlData) {
+                    $scope.subtopicEditorIsShown = true;
+                    $scope.htmlData = previewHtmlData;
+                  };
+
+                  $scope.save = function() {
+                    $uibModalInstance.close({
+                      newTitle: $scope.editableTitle,
+                      newHtmlData: $scope.htmlData
+                    });
+                  };
+
+                  $scope.cancel = function() {
+                    $uibModalInstance.dismiss('cancel');
+                  };
+                }
+              ]
+            });
+
+            modalInstance.result.then(function(newValues) {
+              $scope.subtopicPage = TopicEditorStateService.getSubtopicPage();
+              var newTitle = newValues.newTitle;
+              var newHtmlData = newValues.newHtmlData;
+
+              if ((newTitle !== subtopic.getTitle()) && (newTitle !== '')) {
+                TopicUpdateService.setSubtopicTitle(
+                  $scope.topic, subtopic.getId(), newTitle);
+              }
+              if (newHtmlData !== $scope.subtopicPage.getHtmlData()) {
+                TopicUpdateService.setSubtopicPageHtmlData(
+                  $scope.subtopicPage, subtopic.getId(), newHtmlData);
+                TopicEditorStateService.setSubtopicPage($scope.subtopicPage);
+              }
+            });
           };
 
-          $scope.closeSubtopicTitleEditor = function() {
-            $scope.subtopicTitleEditorIsShown = false;
+          $scope.getSkillEditorUrl = function(subtopicIndex, skillIndex) {
+            var skillId;
+            if (subtopicIndex === null) {
+              skillId = $scope.uncategorizedSkillIds[skillIndex];
+            } else {
+              skillId = $scope.subtopics[
+                subtopicIndex].getSkillIds()[skillIndex];
+            }
+            return '/skill_editor/' + skillId;
           };
 
-          $scope.openPreviewSubtopicPage = function(htmlData) {
-            $scope.subtopicEditorIsShown = false;
-            $scope.htmlData = htmlData;
+          $scope.startMoveSkill = function(oldSubtopicIndex, index) {
```

Comment:
Any way you can pass the ID directly, like you do in endMoveSkill? Having to deal with indexing here seems clumsy.

Reply:
Done.